File: packages/core/src/summarizer/Compressor.ts
/**
 * 压缩器实现
 * 80/20 压缩策略 + 决策骨架提取
 */

import { Message, Context, DecisionSkeleton, getMessageText } from '../hooks/types.js';
import { ICliAdapter } from '../adapters/types.js';
import { HookManager } from '../hooks/HookManager.js';
import {
  ICompressor,
  CompressionConfig,
  CompressionResult,
  SummaryAgentConfig,
  DEFAULT_COMPRESSION_CONFIG,
} from './types.js';
import { TokenCounter } from './TokenCounter.js';

export class Compressor implements ICompressor {
  private tokenCounter: TokenCounter;
  private summaryAdapter?: ICliAdapter;
  private hookManager?: HookManager;
  private config: CompressionConfig;

  constructor(
    summaryAdapter?: ICliAdapter,
    hookManager?: HookManager,
    config?: Partial<CompressionConfig>
  ) {
    this.tokenCounter = new TokenCounter();
    this.summaryAdapter = summaryAdapter;
    this.hookManager = hookManager;
    this.config = { ...DEFAULT_COMPRESSION_CONFIG, ...config };
  }

  async compress(
    context: Context,
    config?: Partial<CompressionConfig>
  ): Promise<CompressionResult> {
    const mergedConfig = { ...this.config, ...config };
    const { messages, tokenCount: originalTokens } = context;

    // 触发 hook_before_compress
    let decisionSkeleton: DecisionSkeleton | undefined;
    if (this.hookManager && mergedConfig.extractDecisionSkeleton) {
      decisionSkeleton = await this.hookManager.hook_before_compress(context);
    } else if (mergedConfig.extractDecisionSkeleton) {
      decisionSkeleton = await this.extractSkeleton(messages);
    }

    // 80/20 压缩策略
    const preservedMessages = this.applyCompressionStrategy(messages, mergedConfig);

    // 生成摘要（如果有 Summary Agent）
    let summary: string | undefined;
    if (this.summaryAdapter && messages.length > preservedMessages.length) {
      const messagesToSummarize = messages.slice(
        0,
        messages.length - mergedConfig.preserveRecentMessages
      );
      summary = await this.generateSummary(messagesToSummarize);
    }

    // 计算压缩后的 token 数
    const compressedTokens = this.tokenCounter.countMessages(preservedMessages).total;

    return {
      originalTokens,
      compressedTokens,
      compressionRatio: compressedTokens / originalTokens,
      preservedMessages,
      summary,
      decisionSkeleton,
    };
  }

  async extractSkeleton(messages: Message[]): Promise<DecisionSkeleton> {
    const entities: string[] = [];
    const decisions: string[] = [];
    const relations: Array<{ from: string; to: string; type: string }> = [];

    // 实体提取模式
    const entityPatterns = [
      // 大写开头的词（类名、专有名词）
      /\b[A-Z][a-zA-Z]{2,}\b/g,
      // 驼峰命名（变量名、函数名）
      /\b[a-z]+(?:[A-Z][a-z]+)+\b/g,
      // 下划线命名
      /\b[a-z]+(?:_[a-z]+)+\b/g,
      // 文件路径
      /(?:[\w-]+\/)+[\w-]+\.\w+/g,
      // 技术术语（带连字符）
      /\b[A-Za-z]+-[A-Za-z]+(?:-[A-Za-z]+)*\b/g,
    ];

    // 决策关键词（扩展）
    const decisionKeywords = [
      // 英文
      'decide', 'decided', 'decision',
      'choose', 'chose', 'chosen', 'choice',
      'select', 'selected', 'selection',
      'implement', 'implemented', 'implementation',
      'use', 'using', 'used',
      'adopt', 'adopted', 'adoption',
      'prefer', 'preferred',
      'recommend', 'recommended',
      'should', 'must', 'will',
      'going to', 'plan to',
      // 中文
      '决定', '选择', '采用', '实现', '使用',
      '推荐', '建议', '应该', '必须', '计划',
    ];

    // 关系关键词
    const relationPatterns = [
      { pattern: /(\w+)\s+(?:extends|inherits from|is a)\s+(\w+)/gi, type: 'extends' },
      { pattern: /(\w+)\s+(?:implements|realizes)\s+(\w+)/gi, type: 'implements' },
      { pattern: /(\w+)\s+(?:uses|depends on|requires)\s+(\w+)/gi, type: 'uses' },
      { pattern: /(\w+)\s+(?:contains|has|includes)\s+(\w+)/gi, type: 'contains' },
      { pattern: /(\w+)\s+(?:calls|invokes)\s+(\w+)/gi, type: 'calls' },
      { pattern: /(\w+)\s+(?:returns|produces)\s+(\w+)/gi, type: 'returns' },
    ];

    // 排除的常见词
    const excludeWords = new Set([
      'The', 'This', 'That', 'These', 'Those', 'What', 'When', 'Where', 'Which', 'Who',
      'How', 'Why', 'Can', 'Could', 'Would', 'Should', 'Will', 'May', 'Might', 'Must',
      'Have', 'Has', 'Had', 'Been', 'Being', 'Are', 'Were', 'Was', 'Is', 'Do', 'Does',
      'Did', 'Done', 'Get', 'Got', 'Set', 'Let', 'Put', 'Make', 'Made', 'Take', 'Took',
      'Give', 'Gave', 'Come', 'Came', 'Go', 'Went', 'See', 'Saw', 'Know', 'Knew',
      'Think', 'Thought', 'Say', 'Said', 'Tell', 'Told', 'Ask', 'Asked', 'Use', 'Used',
      'Find', 'Found', 'Want', 'Need', 'Try', 'Tried', 'Keep', 'Kept', 'Start', 'Stop',
      'True', 'False', 'Null', 'Undefined', 'None', 'Yes', 'No', 'Not', 'And', 'Or',
      'But', 'For', 'With', 'From', 'Into', 'About', 'After', 'Before', 'Between',
    ]);

    for (const msg of messages) {
      const content = getMessageText(msg.content);

      // 提取实体
      for (const pattern of entityPatterns) {
        const matches = content.match(pattern);
        if (matches) {
          for (const entity of matches) {
            if (
              entity.length >= 3 &&
              !excludeWords.has(entity) &&
              !entities.includes(entity)
            ) {
              entities.push(entity);
            }
          }
        }
      }

      // 提取决策
      const sentences = content.split(/[.。!！?？\n]/);
      for (const sentence of sentences) {
        const trimmed = sentence.trim();
        if (trimmed.length < 10) continue;

        const lowerSentence = trimmed.toLowerCase();
        const hasDecisionKeyword = decisionKeywords.some((kw) => lowerSentence.includes(kw));

        if (hasDecisionKeyword && !decisions.includes(trimmed)) {
          // 提取决策的核心部分（限制长度）
          const decision = trimmed.length > 200 ? trimmed.slice(0, 200) + '...' : trimmed;
          decisions.push(decision);
        }
      }

      // 提取关系
      for (const { pattern, type } of relationPatterns) {
        let match;
        const regex = new RegExp(pattern.source, pattern.flags);
        while ((match = regex.exec(content)) !== null) {
          const from = match[1];
          const to = match[2];
          if (
            from !== to &&
            !relations.some((r) => r.from === from && r.to === to && r.type === type)
          ) {
            relations.push({ from, to, type });
          }
        }
      }
    }

    // 基于实体共现补充关系
    const entitySet = new Set(entities.slice(0, 30));
    for (const msg of messages) {
      const content = getMessageText(msg.content);
      const foundEntities = entities.filter((e) => content.includes(e));
      if (foundEntities.length >= 2) {
        for (let i = 0; i < foundEntities.length - 1; i++) {
          for (let j = i + 1; j < foundEntities.length; j++) {
            const from = foundEntities[i];
            const to = foundEntities[j];
            if (!relations.some((r) => r.from === from && r.to === to)) {
              relations.push({ from, to, type: 'related_to' });
            }
          }
        }
      }
    }

    return {
      entities: entities.slice(0, 30),
      decisions: decisions.slice(0, 15),
      relations: relations.slice(0, 25),
    };
  }

  async generateSummary(messages: Message[], config?: SummaryAgentConfig): Promise<string> {
    if (!this.summaryAdapter) {
      return this.generateLocalSummary(messages);
    }

    const prompt = this.buildSummaryPrompt(messages, config);

    try {
      const response = await this.summaryAdapter.send(prompt, {
        maxTokens: config?.maxSummaryTokens ?? 500,
      });
      return response.content;
    } catch {
      return this.generateLocalSummary(messages);
    }
  }

  // ==================== 私有方法 ====================

  private applyCompressionStrategy(messages: Message[], config: CompressionConfig): Message[] {
    const preserved: Message[] = [];
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const tokenCount = this.tokenCounter.countMessages(messages);

    // 1. 保留系统提示
    if (config.preserveSystemPrompt) {
      const systemMessages = messages.filter((m) => m.role === 'system');
      preserved.push(...systemMessages);
    }

    // 2. 保留最近的消息
    const recentMessages = messages
      .filter((m) => m.role !== 'system')
      .slice(-config.preserveRecentMessages);
    preserved.push(...recentMessages);

    // 3. 80/20 策略：保留 20% 最重要的历史消息
    const historicalMessages = messages
      .filter((m) => m.role !== 'system')
      .slice(0, -config.preserveRecentMessages);

    if (historicalMessages.length > 0) {
      const targetHistoricalTokens = Math.floor(
        config.maxTokens * config.targetRatio - this.tokenCounter.countMessages(preserved).total
      );

      if (targetHistoricalTokens > 0) {
        // 按重要性排序历史消息
        const scored = historicalMessages.map((msg, idx) => ({
          msg,
          idx,
          score: this.calculateImportance(msg, idx, historicalMessages.length),
        }));

        scored.sort((a, b) => b.score - a.score);

        let currentTokens = 0;
        const selectedHistorical: Message[] = [];

        for (const { msg } of scored) {
          const msgTokens = this.tokenCounter.count(getMessageText(msg.content));
          if (currentTokens + msgTokens <= targetHistoricalTokens) {
            selectedHistorical.push(msg);
            currentTokens += msgTokens;
          }
        }

        // 按原始顺序排列
        selectedHistorical.sort(
          (a, b) => historicalMessages.indexOf(a) - historicalMessages.indexOf(b)
        );

        // 插入到系统消息之后、最近消息之前
        const systemCount = preserved.filter((m) => m.role === 'system').length;
        preserved.splice(systemCount, 0, ...selectedHistorical);
      }
    }

    return preserved;
  }

  private calculateImportance(msg: Message, index: number, totalMessages: number): number {
    let score = 0;
    const content = getMessageText(msg.content);
    const lowerContent = content.toLowerCase();

    // 1. 长度因素（较长的消息可能更重要，但有上限）
    const lengthScore = Math.min(content.length / 100, 10);
    score += lengthScore;

    // 2. 角色因素
    if (msg.role === 'assistant') {
      score += 5; // assistant 消息通常包含更多信息
    } else if (msg.role === 'system') {
      score += 10; // 系统消息最重要
    }

    // 3. 位置因素（较新的消息更重要）
    const positionScore = (index / totalMessages) * 5;
    score += positionScore;

    // 4. 关键词因素（扩展）
    const keywordScores: Record<string, number> = {
      // 高优先级关键词
      'critical': 5, 'important': 4, 'must': 4, 'required': 4,
      'error': 4, 'bug': 4, 'fix': 3, 'issue': 3,
      'security': 5, 'vulnerability': 5,
      'breaking': 5, 'deprecated': 4,
      // 中优先级关键词
      'should': 2, 'recommend': 2, 'suggest': 2,
      'warning': 3, 'caution': 3,
      'implement': 2, 'feature': 2, 'enhancement': 2,
      // 中文关键词
      '重要': 4, '关键': 4, '必须': 4, '错误': 4, '修复': 3,
      '安全': 5, '漏洞': 5, '问题': 3, '建议': 2,
    };

    for (const [keyword, keywordScore] of Object.entries(keywordScores)) {
      if (lowerContent.includes(keyword)) {
        score += keywordScore;
      }
    }

    // 5. 代码块因素（包含代码的消息可能更重要）
    const codeBlockCount = (content.match(/```/g) || []).length / 2;
    score += Math.min(codeBlockCount * 2, 6);

    // 6. 结构化内容因素（列表、标题等）
    const hasStructure =
      content.includes('- ') ||
      content.includes('* ') ||
      content.includes('1. ') ||
      content.match(/^#+\s/m);
    if (hasStructure) {
      score += 2;
    }

    // 7. 问答因素（问题和回答通常成对重要）
    if (content.includes('?') || content.includes('？')) {
      score += 1;
    }

    return score;
  }

  private generateLocalSummary(messages: Message[]): string {
    const userMessages = messages.filter((m) => m.role === 'user');
    const assistantMessages = messages.filter((m) => m.role === 'assistant');

    // 提取主题（改进：提取更有意义的内容）
    const topics: string[] = [];
    const keyPhrases: string[] = [];

    for (const msg of userMessages.slice(-5)) {
      const content = getMessageText(msg.content);
      const firstSentence = content.split(/[.。!！?？\n]/)[0]?.trim();
      if (firstSentence && firstSentence.length >= 5 && firstSentence.length <= 100) {
        topics.push(firstSentence);
      } else if (content.length > 0) {
        const topic = content.slice(0, 80).replace(/\n/g, ' ').trim();
        if (topic) topics.push(topic + (content.length > 80 ? '...' : ''));
      }
    }

    // 从 assistant 消息中提取关键短语
    for (const msg of assistantMessages.slice(-3)) {
      // 提取包含关键动作的短语
      const actionPatterns = [
        /(?:I |I'll |I've |We |We'll |We've )([^.。!！?？]{10,60})/gi,
        /(?:created|implemented|fixed|added|updated|modified|removed)\s+([^.。!！?？]{5,50})/gi,
      ];

      for (const pattern of actionPatterns) {
        const matches = getMessageText(msg.content).match(pattern);
        if (matches) {
          for (const match of matches.slice(0, 2)) {
            const phrase = match.trim();
            if (phrase.length >= 10 && !keyPhrases.includes(phrase)) {
              keyPhrases.push(phrase);
            }
          }
        }
      }
    }

    // 构建摘要
    const parts: string[] = [];

    parts.push(`Conversation summary (${messages.length} messages):`);

    if (topics.length > 0) {
      parts.push(`Topics discussed: ${topics.slice(0, 3).join('; ')}`);
    }

    if (keyPhrases.length > 0) {
      parts.push(`Key actions: ${keyPhrases.slice(0, 3).join('; ')}`);
    }

    parts.push(
      `Statistics: ${userMessages.length} user messages, ${assistantMessages.length} assistant responses.`
    );

    return parts.join(' ');
  }

  private buildSummaryPrompt(messages: Message[], config?: SummaryAgentConfig): string {
    const content = messages.map((m) => `[${m.role}]: ${getMessageText(m.content).slice(0, 500)}`).join('\n\n');

    let prompt = `Summarize the following conversation concisely:\n\n${content}\n\n`;

    if (config?.includeEntities) {
      prompt += 'Include key entities mentioned.\n';
    }
    if (config?.includeDecisions) {
      prompt += 'Include important decisions made.\n';
    }
    if (config?.includeRelations) {
      prompt += 'Include relationships between concepts.\n';
    }

    prompt += '\nProvide a concise summary:';

    return prompt;
  }
}

File: backend/internal/summarizer/compressor.go
package summarizer

import (
	"context"
	"fmt"
	"log"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/codeflow/backend/internal/adapters"
	backendhooks "github.com/codeflow/backend/internal/hooks"
)

// Compressor 压缩器实现
type Compressor struct {
	tokenCounter   *TokenCounter
	summaryAdapter adapters.ICliAdapter
	config         CompressionConfig
}

// NewCompressor 创建压缩器
func NewCompressor(summaryAdapter adapters.ICliAdapter, config *CompressionConfig) *Compressor {
	c := &Compressor{
		tokenCounter:   NewTokenCounter(nil),
		summaryAdapter: summaryAdapter,
		config:         DefaultCompressionConfig,
	}

	if config != nil {
		c.config = *config
	}

	return c
}

// Compress 压缩上下文
func (c *Compressor) Compress(ctx Context, config *CompressionConfig) (*CompressionResult, error) {
	return c.CompressContext(context.Background(), ctx, config)
}

// CompressContext 使用调用方上下文压缩上下文。
func (c *Compressor) CompressContext(parent context.Context, ctx Context, config *CompressionConfig) (*CompressionResult, error) {
	if parent == nil {
		parent = context.Background()
	}
	ctx = notifyBeforeCompressHook(parent, ctx)
	mergedConfig := c.config
	if config != nil {
		if config.MaxTokens > 0 {
			mergedConfig.MaxTokens = config.MaxTokens
		}
		if config.TargetRatio > 0 {
			mergedConfig.TargetRatio = config.TargetRatio
		}
		mergedConfig.PreserveSystemPrompt = config.PreserveSystemPrompt
		if config.PreserveRecentMessages > 0 {
			mergedConfig.PreserveRecentMessages = config.PreserveRecentMessages
		}
		mergedConfig.ExtractDecisionSkeleton = config.ExtractDecisionSkeleton
	}

	originalTokens := ctx.TokenCount
	if originalTokens == 0 {
		originalTokens = c.tokenCounter.CountMessages(ctx.Messages).Total
	}

	// 提取决策骨架
	var skeleton *DecisionSkeleton
	if mergedConfig.ExtractDecisionSkeleton {
		var err error
		skeleton, err = c.ExtractSkeleton(ctx.Messages)
		if err != nil {
			skeleton = nil
		}
	}

	// 80/20 压缩策略
	preservedMessages := c.applyCompressionStrategy(ctx.Messages, mergedConfig)

	// 生成摘要
	var summary string
	if c.summaryAdapter != nil && len(ctx.Messages) > len(preservedMessages) {
		messagesToSummarize := ctx.Messages[:len(ctx.Messages)-mergedConfig.PreserveRecentMessages]
		if len(messagesToSummarize) > 0 {
			var err error
			summary, err = c.GenerateSummaryContext(parent, messagesToSummarize, nil)
			if err != nil {
				summary = c.generateLocalSummary(messagesToSummarize)
			}
		}
	}

	compressedTokens := c.tokenCounter.CountMessages(preservedMessages).Total

	return &CompressionResult{
		OriginalTokens:    originalTokens,
		CompressedTokens:  compressedTokens,
		CompressionRatio:  float64(compressedTokens) / float64(originalTokens),
		PreservedMessages: preservedMessages,
		Summary:           summary,
		DecisionSkeleton:  skeleton,
	}, nil
}

func notifyBeforeCompressHook(ctx context.Context, payload Context) Context {
	if !backendhooks.HasHookManager() {
		return payload
	}
	result, err := backendhooks.GetHookManager().Trigger(ctx, backendhooks.HookBeforeCompress, payload)
	if err != nil {
		log.Printf("[WARN] summarizer before-compress hook failed: err=%v", err)
		return payload
	}
	switch converted := result.(type) {
	case Context:
		return converted
	case *Context:
		if converted != nil {
			return *converted
		}
	}
	return payload
}

// ExtractSkeleton 提取决策骨架
func (c *Compressor) ExtractSkeleton(messages []adapters.Message) (*DecisionSkeleton, error) {
	entities := make([]string, 0)
	decisions := make([]string, 0)
	relations := make([]Relation, 0)

	entitySet := make(map[string]bool)
	decisionSet := make(map[string]bool)

	// 实体提取正则（大写开头的词）
	entityRegex := regexp.MustCompile(`\b[A-Z][a-zA-Z]+\b`)

	// 决策关键词
	decisionKeywords := []string{
		"decide", "choose", "select", "implement", "use", "adopt",
		"决定", "选择", "采用", "实现",
	}

	for _, msg := range messages {
		content := msg.Content

		// 提取实体
		entityMatches := entityRegex.FindAllString(content, -1)
		for _, entity := range entityMatches {
			if !entitySet[entity] && len(entity) > 2 {
				entitySet[entity] = true
				entities = append(entities, entity)
			}
		}

		// 提取决策
		sentences := splitSentences(content)
		for _, sentence := range sentences {
			lowerSentence := strings.ToLower(sentence)
			for _, kw := range decisionKeywords {
				if strings.Contains(lowerSentence, kw) {
					trimmed := strings.TrimSpace(sentence)
					if trimmed != "" && !decisionSet[trimmed] && len(trimmed) > 10 {
						decisionSet[trimmed] = true
						decisions = append(decisions, trimmed)
					}
					break
				}
			}
		}
	}

	// 提取关系（基于实体共现）
	entityList := entities
	for i := 0; i < len(entityList)-1 && i < 20; i++ {
		for j := i + 1; j < len(entityList) && j < 20; j++ {
			e1, e2 := entityList[i], entityList[j]
			for _, msg := range messages {
				if strings.Contains(msg.Content, e1) && strings.Contains(msg.Content, e2) {
					relations = append(relations, Relation{
						From: e1,
						To:   e2,
						Type: "related_to",
					})
					break
				}
			}
		}
	}

	// 限制数量
	if len(entities) > 20 {
		entities = entities[:20]
	}
	if len(decisions) > 10 {
		decisions = decisions[:10]
	}
	if len(relations) > 15 {
		relations = relations[:15]
	}

	return &DecisionSkeleton{
		Entities:  entities,
		Decisions: decisions,
		Relations: relations,
	}, nil
}

// GenerateSummary 生成总结
func (c *Compressor) GenerateSummary(messages []adapters.Message, config *SummaryAgentConfig) (string, error) {
	return c.GenerateSummaryContext(context.Background(), messages, config)
}

// GenerateSummaryContext 使用调用方上下文生成总结。
func (c *Compressor) GenerateSummaryContext(ctx context.Context, messages []adapters.Message, config *SummaryAgentConfig) (string, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	if c.summaryAdapter == nil {
		return c.generateLocalSummary(messages), nil
	}

	prompt := c.buildSummaryPrompt(messages, config)

	maxTokens := 500
	if config != nil && config.MaxSummaryTokens > 0 {
		maxTokens = config.MaxSummaryTokens
	}

	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()

	response, err := c.summaryAdapter.Send(ctx, prompt, &adapters.SendOptions{
		MaxTokens: maxTokens,
	})
	if err != nil {
		return c.generateLocalSummary(messages), nil
	}

	return response.Content, nil
}

// applyCompressionStrategy 应用80/20压缩策略
func (c *Compressor) applyCompressionStrategy(messages []adapters.Message, config CompressionConfig) []adapters.Message {
	preserved := make([]adapters.Message, 0)

	// 1. 保留系统提示
	systemMessages := make([]adapters.Message, 0)
	if config.PreserveSystemPrompt {
		for _, m := range messages {
			if m.Role == adapters.RoleSystem {
				systemMessages = append(systemMessages, m)
			}
		}
		preserved = append(preserved, systemMessages...)
	}

	// 2. 保留最近的消息
	nonSystemMessages := make([]adapters.Message, 0)
	for _, m := range messages {
		if m.Role != adapters.RoleSystem {
			nonSystemMessages = append(nonSystemMessages, m)
		}
	}

	recentCount := config.PreserveRecentMessages
	if recentCount > len(nonSystemMessages) {
		recentCount = len(nonSystemMessages)
	}

	recentMessages := nonSystemMessages[len(nonSystemMessages)-recentCount:]
	historicalMessages := nonSystemMessages[:len(nonSystemMessages)-recentCount]

	// 3. 80/20策略：保留20%最重要的历史消息
	if len(historicalMessages) > 0 {
		preservedTokens := c.tokenCounter.CountMessages(preserved).Total
		recentTokens := c.tokenCounter.CountMessages(recentMessages).Total
		targetHistoricalTokens := int(float64(config.MaxTokens)*config.TargetRatio) - preservedTokens - recentTokens

		if targetHistoricalTokens > 0 {
			// 按重要性评分
			type scoredMessage struct {
				msg   adapters.Message
				idx   int
				score float64
			}

			scored := make([]scoredMessage, 0, len(historicalMessages))
			for idx, msg := range historicalMessages {
				scored = append(scored, scoredMessage{
					msg:   msg,
					idx:   idx,
					score: c.calculateImportance(msg, idx, len(historicalMessages)),
				})
			}

			// 按分数降序排列
			sort.Slice(scored, func(i, j int) bool {
				return scored[i].score > scored[j].score
			})

			// 选择高分消息
			currentTokens := 0
			selectedHistorical := make([]scoredMessage, 0)

			for _, sm := range scored {
				msgTokens := c.tokenCounter.Count(sm.msg.Content)
				if currentTokens+msgTokens <= targetHistoricalTokens {
					selectedHistorical = append(selectedHistorical, sm)
					currentTokens += msgTokens
				}
			}

			// 按原始顺序排列
			sort.Slice(selectedHistorical, func(i, j int) bool {
				return selectedHistorical[i].idx < selectedHistorical[j].idx
			})

			// 插入到系统消息之后
			for _, sm := range selectedHistorical {
				preserved = append(preserved, sm.msg)
			}
		}
	}

	// 添加最近消息
	preserved = append(preserved, recentMessages...)

	return preserved
}

// calculateImportance 计算消息重要性
func (c *Compressor) calculateImportance(msg adapters.Message, index, totalMessages int) float64 {
	score := 0.0

	// 长度因素
	lengthScore := float64(len(msg.Content)) / 100.0
	if lengthScore > 10 {
		lengthScore = 10
	}
	score += lengthScore

	// 角色因素
	if msg.Role == adapters.RoleAssistant {
		score += 5
	}

	// 位置因素（较新的更重要）
	if totalMessages > 0 {
		score += (float64(index) / float64(totalMessages)) * 3
	}

	// 关键词因素
	importantKeywords := []string{
		"important", "critical", "must", "should", "error", "bug", "fix",
		"重要", "关键", "必须", "错误", "修复",
	}
	lowerContent := strings.ToLower(msg.Content)
	for _, kw := range importantKeywords {
		if strings.Contains(lowerContent, kw) {
			score += 2
		}
	}

	return score
}

// generateLocalSummary 生成本地摘要
func (c *Compressor) generateLocalSummary(messages []adapters.Message) string {
	userMessages := make([]adapters.Message, 0)
	assistantMessages := make([]adapters.Message, 0)

	for _, msg := range messages {
		switch msg.Role {
		case adapters.RoleUser:
			userMessages = append(userMessages, msg)
		case adapters.RoleAssistant:
			assistantMessages = append(assistantMessages, msg)
		}
	}

	topics := make([]string, 0)
	count := 5
	if len(userMessages) < count {
		count = len(userMessages)
	}

	for _, msg := range userMessages[len(userMessages)-count:] {
		topic := msg.Content
		if len(topic) > 50 {
			topic = topic[:50]
		}
		topic = strings.ReplaceAll(topic, "\n", " ")
		topics = append(topics, topic)
	}

	return fmt.Sprintf("Previous conversation covered %d messages. User topics: %s. Assistant provided %d responses.",
		len(messages), strings.Join(topics, "; "), len(assistantMessages))
}

// buildSummaryPrompt 构建摘要提示词
func (c *Compressor) buildSummaryPrompt(messages []adapters.Message, config *SummaryAgentConfig) string {
	var sb strings.Builder
	sb.WriteString("Summarize the following conversation concisely:\n\n")

	for _, m := range messages {
		content := m.Content
		if len(content) > 500 {
			content = content[:500]
		}
		sb.WriteString(fmt.Sprintf("[%s]: %s\n\n", m.Role, content))
	}

	if config != nil {
		if config.IncludeEntities {
			sb.WriteString("Include key entities mentioned.\n")
		}
		if config.IncludeDecisions {
			sb.WriteString("Include important decisions made.\n")
		}
		if config.IncludeRelations {
			sb.WriteString("Include relationships between concepts.\n")
		}
	}

	sb.WriteString("\nProvide a concise summary:")
	return sb.String()
}

// splitSentences 分割句子
func splitSentences(text string) []string {
	// 按中英文句子结束符分割
	sentenceRegex := regexp.MustCompile(`[.。!！?？]+`)
	parts := sentenceRegex.Split(text, -1)

	sentences := make([]string, 0, len(parts))
	for _, p := range parts {
		trimmed := strings.TrimSpace(p)
		if trimmed != "" {
			sentences = append(sentences, trimmed)
		}
	}
	return sentences
}

// InMemorySummaryHistory 内存总结历史存储
type InMemorySummaryHistory struct {
	histories map[string]*SummaryHistory
}

// NewInMemorySummaryHistory 创建内存总结历史存储
func NewInMemorySummaryHistory() *InMemorySummaryHistory {
	return &InMemorySummaryHistory{
		histories: make(map[string]*SummaryHistory),
	}
}

// Save 保存历史
func (h *InMemorySummaryHistory) Save(history *SummaryHistory) error {
	h.histories[history.ID] = history
	return nil
}

// Load 加载历史
func (h *InMemorySummaryHistory) Load(id string) (*SummaryHistory, error) {
	if hist, ok := h.histories[id]; ok {
		return hist, nil
	}
	return nil, nil
}

// LoadBySession 按会话加载历史
func (h *InMemorySummaryHistory) LoadBySession(sessionID string) ([]*SummaryHistory, error) {
	result := make([]*SummaryHistory, 0)
	for _, hist := range h.histories {
		if hist.SessionID == sessionID {
			result = append(result, hist)
		}
	}
	return result, nil
}

// Delete 删除历史
func (h *InMemorySummaryHistory) Delete(id string) error {
	delete(h.histories, id)
	return nil
}

// List 列出所有历史
func (h *InMemorySummaryHistory) List() ([]*SummaryHistory, error) {
	result := make([]*SummaryHistory, 0, len(h.histories))
	for _, hist := range h.histories {
		result = append(result, hist)
	}
	return result, nil
}
